### PR TITLE
selinux: make selinux access checks based on unit file contexts read at unit load time

### PR DESCRIPTION
**Pull request by poettering**
As per discussion:

https://github.com/systemd/systemd/pull/10023#issuecomment-649325934

### CODE REVIEW

**Comment by poettering** [comment]
Tested lightly only.

**Comment by cgzones** [comment]
Just for completeness: The cached context might be invalid by the time of an access check, e.g. because the used type got removed. For stable policies (like Fedora or Refpolicy) this should not be an issue, as they rarely remove a type and if they do they normally add a backwards compatibility alias.
(So checking the context against `security_check_context(3)` and using `/sys/fs/selinux/initial_contexts/unlabeled` as fallback is probably an unnecessary overhead.)

**Comment by bluca** [comment]
```
../src/src/core/selinux-access.c:276:5: error: conflicting types for ‘mac_selinux_access_check_internal’
  276 | int mac_selinux_access_check_internal(
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../src/src/core/selinux-access.c:3:
../src/src/core/selinux-access.h:8:5: note: previous declaration of ‘mac_selinux_access_check_internal’ was here
    8 | int mac_selinux_access_check_internal(sd_bus_message *message, const char *unit_path, const char *unit_label, const char *permission, const char *function, sd_bus_error *error);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by poettering** [comment]
Rebased, and hopefully build issue addressed. PTAL

File: src/core/selinux-access.c [hunk 1]
```diff
@@ -280,18 +267,19 @@ int mac_selinux_access_check_internal(
...
 
 int mac_selinux_access_check_internal(
                 sd_bus_message *message,
-                const char *path,
+                const char *unit_path,
+                const char *unit_label,
                 const char *permission,
                 const char *function,
-                sd_bus_error *error) {
+                sd_bus_error *error);
```
Comment (bluca) [inline src/core/selinux-access.c]: this doesn't look right?

**Comment by poettering** [comment]
fixed.

**Review by bluca — APPROVED** [review]

**Comment by pebenito** [comment]
Looks good to me.

File: src/core/load-fragment.c [hunk 2]
```diff
@@ -6141,7 +6143,23 @@ int unit_load_fragment(Unit *u) {
...
+                        r = fgetfilecon_raw(fileno(f), &selcon);
+                        if (r < 0)
+                                log_unit_warning_errno(u, r, "Failed to read SELinux context of '%s', ignoring: %m", fragment);
+
+                        r = free_and_strdup(&u->access_selinux_context, selcon);
+                        if (r < 0)
+                                return r;
+#else
+                        u->access_selinux_context = mfree(u->access_selinux_context);
+#endif
```
Comment (cgzones) [inline src/core/load-fragment.c:6161]: Suggestion:

```diff
#if HAVE_SELINUX
                        if (mac_selinux_use()) {
                            _cleanup_freecon_ char *selcon = NULL;

                            /* Cache the SELinux context of the unit file here. We'll make use of when checking access permissions to loaded units */
                            r = fgetfilecon_raw(fileno(f), &selcon);
                            if (r < 0) {
                               log_unit_warning_errno(u, r, "Failed to read SELinux context of '%s', ignoring: %m", fragment);

                            r = free_and_strdup(&u->access_selinux_context, selcon);
                            if (r < 0)
                                    return r;
                        } else
#endif
                        {
                            u->access_selinux_context = mfree(u->access_selinux_context);
                        }
```

**Comment by poettering** [comment]
pushed new version. Only change is along the lines of what @cgzones suggested.

**Comment by poettering** [comment]
CI failures appear unrelated (fwupd failed, 2 canonical hosts didn't resolve, and dissect timed-out)

merging.

**Pull request merged**